MBBSEmu/DOS/Interrupts/Int21h.cs

```diff
@@ -90,443 +90,475 @@ public void Handle()
+                BX = file handle
+                CX = number of bytes to write
+                DS:DX -> buffer
+
+                Return: CF set on error
+                    AX = error code
+
+                CF clear if successful
+                    AX = number of bytes written
+
+                Note: if CX is zero, no data is written, and the file is truncated or extended
+                    to the current position
+            */
+            var fileHandle = _registers.BX;
+            var numberOfBytes = _registers.CX;
+            var bufferPointer = new FarPtr(_registers.DS, _registers.DX);
+
+            var dataToWrite = _memory.GetArray(bufferPointer, numberOfBytes);
+            char[] toWrite = new char[numberOfBytes];
+            Encoding.ASCII.GetChars(dataToWrite, toWrite.AsSpan());
+
+            switch (fileHandle)
+            {
+                case (ushort)FileHandle.STDOUT:
+                    _stdout.Write(toWrite);
+                    break;
+                case (ushort)FileHandle.STDERR:
+                    _stderr.Write(toWrite);
+                    break;
                 default:
-                    throw new ArgumentOutOfRangeException($"Unsupported INT 21h Function: 0x{_registers.AH:X2}");
+                    _logger.Warn($"Write to file not yet supported");
+                    _registers.F = _registers.F.SetFlag((ushort)EnumFlags.CF);
+                    _registers.AX = (ushort)DOSErrorCode.INVALID_HANDLE;
+                    return;
             }
+
+            _registers.F = _registers.F.ClearFlag((ushort)EnumFlags.CF);
+            _registers.AX = numberOfBytes;
+        }
+
+        private void GetCurrentDirectory_0x47()
+        {
+            /*
+                DOS 2+ - GET CURRENT DIRECTORY
+                DL = drive (0=default, 1=A, etc.)
+                DS:SI points to 64-byte buffer area
+                Return: CF set on error
+                AX = error code
+                Note: the returned path does not include the initial backslash
+            */
+            _memory.SetArray(_registers.DS, _registers.SI, Encoding.ASCII.GetBytes("BBSV6\0"));
+            _registers.DL = DEFAULT_BLOCK_DEVICE;
+            _registers.F = _registers.F.ClearFlag((ushort)EnumFlags.CF);
+        }
+
+        private void AdjustMemoryBlockSize_0x4A()
+        {
+            /*
+                INT 21 - AH = 4Ah DOS 2+ - ADJUST MEMORY BLOCK SIZE (SETBLOCK)
+                ES = Segment address of block to change
+                BX = New size in paragraphs
+                Return: CF set on error
+                AX = error code
+                BX = maximum size possible for the block
+
+                Because MBBSEmu allocates blocks as 0xFFFF in length, we ignore this and proceed
+            */
+
+            var segmentToAdjust = _registers.ES;
+            var newSize = _registers.BX;
+
+            if (_memory is ProtectedModeMemoryCore)
```

Comment:
you can combine the cast and type check here

`_memory is ProtectedModeMemoryCore protectedMemory`

Reply:
nice, forgot about that